pkg/ottl/ottlfuncs/func_set.go

```diff
@@ -38,11 +64,99 @@ func set[K any](target ottl.Setter[K], value ottl.Getter[K]) ottl.ExprFunc[K] {
 
 		// No fields currently support `null` as a valid type.
 		if val != nil {
+			oldVal, err := target.Get(ctx, tCtx)
+			if err != nil {
+				return nil, err
+			}
+
+			valueMissing := oldVal == nil || isDefaultValue(oldVal)
+			switch conflictStrategy {
+			case setConflictUpsert:
+				// overwrites if present or create when missing
+			case setConflictUpdate:
+				// update existing field with a new value. If the field does not exist, then no action will be taken
+				if valueMissing {
+					return nil, nil
+				}
+			case setConflictFail:
+				// fail if target field present
+				if !valueMissing {
+					return nil, ErrValueAlreadyPresent
+				}
+			case setConflictInsert:
+				// noop if target field present
+				if !valueMissing {
+					return nil, nil
+				}
+			}
+
 			err = target.Set(ctx, tCtx, val)
 			if err != nil {
 				return nil, err
 			}
 		}
 		return nil, nil
+	}, nil
+}
+
+func isDefaultValue(val any) bool {
```

Comment:
if we have something better for this please let me know. i see following pattern repeated for many use cases